File: iOS/Prosary/Support/SwiftData/PresetEntry.swift
//
//  PresetEntry.swift
//  Prosary
//

import SwiftData
import Foundation

// CloudKit-backed SwiftData stores (see AppServices.modelContainer) impose two constraints on
// every @Model this schema declares, neither of which held before this file was made syncable:
// no unique constraints (CloudKit records have no equivalent) — id is no longer @Attribute(.unique),
// with the store's own application-level id-matching in SwiftDataPresetStore doing the same job the
// database-level constraint used to — and every property needs an inline default value (a value
// only ever assigned inside a custom init doesn't count towards CloudKit's schema validation).
@Model
final class PresetEntry {
  var id: UUID = UUID()
  var name: String = ""
  var isDefault: Bool = false
  var languageCode: String = ""

  // Discriminant — "rosary" / "jesusPrayer" / "custom", plus legacy per-devotion strings
  // ("angelus", "stationsOfTheCross", ...) on rows written before the generic-devotion
  // migration — see `resolvedKind`. Default "rosary" handles existing rows when the schema is
  // extended with this field.
  var kind: String = "rosary"

  // Bundle id for a generic (kind == "custom") devotion, e.g. "trisagion". Nil for every other
  // kind. Optional handles existing rows predating this field.
  var customDevotionId: String? = nil

  // Which bundle variant (alternate step-set) this favorite prays; nil = the bundle's default.
  // Optional handles existing rows predating this field (SwiftData lightweight migration).
  var variantId: String? = nil

  // JSON-encoded [String: String] of the favorite's options.json choices (see
  // Prayer.customOptions). Default empty Data handles existing rows.
  var customOptionsJSON: Data = Data()

  // Rosary-specific fields (populated when kind == "rosary")
  var mysterySelectionMode: MysterySelectionMode = MysterySelectionMode.todaysMysteries
  var specificMysteryGroup: MysteryGroup = MysteryGroup.joyful
  // 1-based index into MysteryCatalog.forGroup(specificMysteryGroup); used only when
  // mysterySelectionMode is .singleMystery. Default 1 handles existing rows.
  var specificMysteryOrder: Int = 1
  var includeApostlesCreed: Bool = true
  var includeOpeningPrayers: Bool = true
  var includeFatimaPrayers: Bool = true
  var eternalRestForDeceased: EternalRestPlacement = EternalRestPlacement.none
  var marianAntiphon: MarianAntiphonOption = MarianAntiphonOption.seasonal
  // Default false handles existing rows.
  var includeClosingIntentions: Bool = false
  var includeStMichaelPrayer: Bool = false
  var includeFinalSignOfCross: Bool = true
  // Raw String for safe lightweight migration of existing SwiftData rows.
  var aramaicSignOfCrossForm: String = AramaicSignOfCrossForm.formA
  // Default false handles existing rows.
  var presenterMode: Bool = false
  // Default "classic" handles existing rows. Stored as the raw string, not the enum type:
  // SwiftData lightweight-migrates a missing String column to its default, but a column of an
  // enum type added to an existing store trips a swift_dynamicCastFailure abort in the
  // generated getter the first time an old row is read (observed 2026-08-13; the older enum
  // columns above predate every shipped store, so they never hit this).
  var mysteryImageStyleRaw: String = MysteryImageStyle.classic.rawValue

  // Jesus Prayer-specific fields (populated when kind == "jesusPrayer")
  var jesusPrayerIsUnbounded: Bool = false
  var jesusPrayerCount: Int = 33

  // Reminders — JSON-encoded [PrayerReminder]. Default empty Data handles existing rows.
  var remindersJSON: Data = Data()

  init(prayer: Prayer) {
    id = prayer.id
    name = prayer.name
    isDefault = prayer.isDefault
    languageCode = prayer.languageCode
    kind = prayer.kind.rawValue
    customDevotionId = prayer.customDevotionId
    variantId = prayer.variantId
    customOptionsJSON = (try? JSONEncoder().encode(prayer.customOptions)) ?? Data()

    mysterySelectionMode = prayer.rosary.mysterySelectionMode
    specificMysteryGroup = prayer.rosary.specificMysteryGroup
    specificMysteryOrder = prayer.rosary.specificMysteryOrder
    includeApostlesCreed = prayer.rosary.includeApostlesCreed
    includeOpeningPrayers = prayer.rosary.includeOpeningPrayers
    includeFatimaPrayers = prayer.rosary.includeFatimaPrayer
    eternalRestForDeceased = prayer.rosary.eternalRestForDeceased
    marianAntiphon = prayer.rosary.marianAntiphon
    includeClosingIntentions = prayer.rosary.includeClosingIntentions
    includeStMichaelPrayer = prayer.rosary.includeStMichaelPrayer
    includeFinalSignOfCross = prayer.rosary.includeFinalSignOfCross
    aramaicSignOfCrossForm = prayer.rosary.aramaicSignOfCrossForm
    presenterMode = prayer.rosary.presenterMode
    mysteryImageStyleRaw = prayer.rosary.mysteryImageStyle.rawValue

    if case .count(let n) = prayer.jesusPrayer.target {
      jesusPrayerIsUnbounded = false
      jesusPrayerCount = n
    } else {
      jesusPrayerIsUnbounded = true
      jesusPrayerCount = 33
    }

    remindersJSON = (try? JSONEncoder().encode(prayer.reminders)) ?? Data()
  }

  func update(from prayer: Prayer) {
    name = prayer.name
    isDefault = prayer.isDefault
    languageCode = prayer.languageCode
    kind = prayer.kind.rawValue
    customDevotionId = prayer.customDevotionId
    variantId = prayer.variantId
    customOptionsJSON = (try? JSONEncoder().encode(prayer.customOptions)) ?? Data()

    mysterySelectionMode = prayer.rosary.mysterySelectionMode
    specificMysteryGroup = prayer.rosary.specificMysteryGroup
    specificMysteryOrder = prayer.rosary.specificMysteryOrder
    includeApostlesCreed = prayer.rosary.includeApostlesCreed
    includeOpeningPrayers = prayer.rosary.includeOpeningPrayers
    includeFatimaPrayers = prayer.rosary.includeFatimaPrayer
    eternalRestForDeceased = prayer.rosary.eternalRestForDeceased
    marianAntiphon = prayer.rosary.marianAntiphon
    includeClosingIntentions = prayer.rosary.includeClosingIntentions
    includeStMichaelPrayer = prayer.rosary.includeStMichaelPrayer
    includeFinalSignOfCross = prayer.rosary.includeFinalSignOfCross
    aramaicSignOfCrossForm = prayer.rosary.aramaicSignOfCrossForm
    presenterMode = prayer.rosary.presenterMode
    mysteryImageStyleRaw = prayer.rosary.mysteryImageStyle.rawValue

    if case .count(let n) = prayer.jesusPrayer.target {
      jesusPrayerIsUnbounded = false
      jesusPrayerCount = n
    } else {
      jesusPrayerIsUnbounded = true
      jesusPrayerCount = 33
    }

    remindersJSON = (try? JSONEncoder().encode(prayer.reminders)) ?? Data()
  }

  /// Maps this row's stored kind to the post-generic-devotions model: an unknown raw kind
  /// string is a legacy per-devotion kind ("angelus", "stationsOfTheCross", ...) whose rawValue
  /// deliberately doubles as its bundle id. Permanent read-time behavior, NOT a one-shot
  /// migration — CloudKit can sync rows written by old app versions in at any time.
  var resolvedKind: (kind: PrayerKind, customDevotionId: String?) {
    if let known = PrayerKind(rawValue: kind) {
      return (known, customDevotionId)
    }
    return (.custom, customDevotionId ?? kind)
  }

  func toPrayer() -> Prayer {
    let jpTarget: JesusPrayerTarget = jesusPrayerIsUnbounded
      ? .unbounded
      : .count(jesusPrayerCount)

    let resolved = resolvedKind
    return Prayer(
      id: id,
      name: name,
      kind: resolved.kind,
      isDefault: isDefault,
      languageCode: languageCode,
      rosary: RosaryOptions(
        mysterySelectionMode: mysterySelectionMode,
        specificMysteryGroup: specificMysteryGroup,
        specificMysteryOrder: specificMysteryOrder,
        includeApostlesCreed: includeApostlesCreed,
        includeOpeningPrayers: includeOpeningPrayers,
        includeFatimaPrayer: includeFatimaPrayers,
        eternalRestForDeceased: eternalRestForDeceased,
        marianAntiphon: marianAntiphon,
        includeClosingIntentions: includeClosingIntentions,
        includeStMichaelPrayer: includeStMichaelPrayer,
        includeFinalSignOfCross: includeFinalSignOfCross,
        aramaicSignOfCrossForm: aramaicSignOfCrossForm,
        presenterMode: presenterMode,
        mysteryImageStyle: MysteryImageStyle(rawValue: mysteryImageStyleRaw) ?? .classic
      ),
      jesusPrayer: JesusPrayerOptions(target: jpTarget),
      customDevotionId: resolved.customDevotionId,
      variantId: variantId,
      customOptions: (try? JSONDecoder().decode([String: String].self, from: customOptionsJSON)) ?? [:],
      reminders: (try? JSONDecoder().decode([PrayerReminder].self, from: remindersJSON)) ?? []
    )
  }
}

File: Windows/Prosary/Persistence/PresetEntry.cs
using System.Text.Json;
using Prosary.Models;
using SQLite;

namespace Prosary.Persistence;

/// <summary>
/// SQLite row for a saved <see cref="Prayer"/> favorite — kept as a distinct storage shape from
/// the domain <see cref="Prayer"/> record, matching iOS's <c>PresetEntry</c>/Android's
/// <c>PresetEntity</c> split. Ported from irosary's <c>RosaryConfig</c> (which this table is a
/// structural descendant of), extended with a <see cref="Kind"/> discriminator, Jesus-Prayer
/// fields, and JSON-encoded reminders — none of which existed on irosary's Rosary-only schema.
/// </summary>
public sealed class PresetEntry
{
    [PrimaryKey]
    public Guid Id { get; set; } = Guid.NewGuid();

    [MaxLength(80), NotNull]
    public string Name { get; set; } = "My Prayer";

    /// <summary>The starred/primary favorite for its <see cref="Kind"/>. At most one per kind.</summary>
    public bool IsDefault { get; set; }

    /// <summary>Empty string means "follow the app-level default" — see <see cref="LanguageCatalog.DefaultSentinel"/>.</summary>
    [NotNull]
    public string LanguageCode { get; set; } = LanguageCatalog.DefaultSentinel;

    public PrayerKind Kind { get; set; } = PrayerKind.Rosary;

    /// <summary>Bundle id for a generic (Kind == Custom) devotion, e.g. "trisagion". Null for
    /// every other kind.</summary>
    public string? CustomDevotionId { get; set; } = null;

    /// <summary>Which bundle variant (alternate step-set) this favorite prays; null = the
    /// bundle's default. CreateTableAsync auto-adds this column to existing databases.</summary>
    public string? VariantId { get; set; } = null;

    // Flattened RosaryOptions (only meaningful when Kind == Rosary).
    public MysterySelectionMode MysterySelectionMode { get; set; } = MysterySelectionMode.TodaysMysteries;
    public MysteryGroup SpecificMysteryGroup { get; set; } = MysteryGroup.Joyful;

    /// <summary>1-based index into MysteryCatalog.ForGroup(SpecificMysteryGroup); used only when
    /// MysterySelectionMode is SingleMystery. Defaults to 1 for existing rows.</summary>
    public int SpecificMysteryOrder { get; set; } = 1;

    public bool IncludeApostlesCreed { get; set; } = true;
    public bool IncludeOpeningPrayers { get; set; } = true;
    public bool IncludeFatimaPrayer { get; set; } = true;
    public EternalRestPlacement EternalRestForDeceased { get; set; } = EternalRestPlacement.None;
    public MarianAntiphonOption MarianAntiphon { get; set; } = MarianAntiphonOption.Seasonal;
    public bool IncludeStMichaelPrayer { get; set; }
    public bool IncludeFinalSignOfCross { get; set; } = true;

    /// <summary>Defaults to false for existing rows.</summary>
    public bool PresenterMode { get; set; }

    // Flattened JesusPrayerOptions (only meaningful when Kind == JesusPrayer).
    public bool JesusPrayerIsUnbounded { get; set; }
    public int JesusPrayerCount { get; set; } = 33;

    /// <summary>JSON-encoded <c>List&lt;PrayerReminder&gt;</c> — matches iOS's <c>remindersJSON: Data</c>
    /// approach rather than a separate reminders table, since it's always loaded with its parent
    /// and never queried independently.</summary>
    [NotNull]
    public string RemindersJson { get; set; } = "[]";

    public Prayer ToPrayer() => new()
    {
        Id = Id,
        Name = Name,
        Kind = Kind,
        IsDefault = IsDefault,
        LanguageCode = LanguageCode,
        CustomDevotionId = CustomDevotionId,
        VariantId = VariantId,
        Rosary = new RosaryOptions
        {
            MysterySelectionMode = MysterySelectionMode,
            SpecificMysteryGroup = SpecificMysteryGroup,
            SpecificMysteryOrder = SpecificMysteryOrder,
            IncludeApostlesCreed = IncludeApostlesCreed,
            IncludeOpeningPrayers = IncludeOpeningPrayers,
            IncludeFatimaPrayer = IncludeFatimaPrayer,
            EternalRestForDeceased = EternalRestForDeceased,
            MarianAntiphon = MarianAntiphon,
            IncludeStMichaelPrayer = IncludeStMichaelPrayer,
            IncludeFinalSignOfCross = IncludeFinalSignOfCross,
            PresenterMode = PresenterMode,
        },
        JesusPrayer = new JesusPrayerOptions
        {
            Target = JesusPrayerIsUnbounded
                ? new JesusPrayerTarget.Unbounded()
                : new JesusPrayerTarget.Count(JesusPrayerCount),
        },
        Reminders = DeserializeReminders(RemindersJson),
    };

    public static PresetEntry FromPrayer(Prayer prayer) => new()
    {
        Id = prayer.Id,
        Name = prayer.Name,
        Kind = prayer.Kind,
        IsDefault = prayer.IsDefault,
        LanguageCode = prayer.LanguageCode,
        CustomDevotionId = prayer.CustomDevotionId,
        VariantId = prayer.VariantId,
        MysterySelectionMode = prayer.Rosary.MysterySelectionMode,
        SpecificMysteryGroup = prayer.Rosary.SpecificMysteryGroup,
        SpecificMysteryOrder = prayer.Rosary.SpecificMysteryOrder,
        IncludeApostlesCreed = prayer.Rosary.IncludeApostlesCreed,
        IncludeOpeningPrayers = prayer.Rosary.IncludeOpeningPrayers,
        IncludeFatimaPrayer = prayer.Rosary.IncludeFatimaPrayer,
        EternalRestForDeceased = prayer.Rosary.EternalRestForDeceased,
        MarianAntiphon = prayer.Rosary.MarianAntiphon,
        IncludeStMichaelPrayer = prayer.Rosary.IncludeStMichaelPrayer,
        IncludeFinalSignOfCross = prayer.Rosary.IncludeFinalSignOfCross,
        PresenterMode = prayer.Rosary.PresenterMode,
        JesusPrayerIsUnbounded = prayer.JesusPrayer.Target is JesusPrayerTarget.Unbounded,
        JesusPrayerCount = prayer.JesusPrayer.Target is JesusPrayerTarget.Count(var n) ? n : 33,
        RemindersJson = JsonSerializer.Serialize(prayer.Reminders),
    };

    private static List<PrayerReminder> DeserializeReminders(string json)
    {
        try
        {
            return JsonSerializer.Deserialize<List<PrayerReminder>>(json) ?? [];
        }
        catch (JsonException)
        {
            return [];
        }
    }
}
